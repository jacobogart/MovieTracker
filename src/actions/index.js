export const addDiscoverMovies = (movies) => ({
  type: 'ADD_DISCOVER_MOVIES',
  movies
});

export const updateUser = (id, name, favorites) => ({
  type: 'UPDATE_USER',
  id, 
  name,
  favorites
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});
