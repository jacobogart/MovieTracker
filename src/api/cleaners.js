import { imageUrl } from '../utils/API-logins';

export const cleanAllMovies = (moviesArray) => {
  const cleanMovies = moviesArray.map(movie => {
    const { 
      id, 
      title, 
      poster_path, 
      overview, 
      release_date, 
      vote_average 
    } = movie;

    return { 
      id, 
      title, 
      posterImg: `${imageUrl}${poster_path}`, 
      overview, 
      releaseDate: release_date, 
      rating: vote_average 
    };
  });

  return cleanMovies;
};

export const cleanPeople = (data) => {
  const director = data.crew.filter(person => person.job === "Director");
  const writer = data.crew.filter(person => person.job === "Writer");
  const cast = data.cast.map(person => {
    let { name, character, profile_path,  cast_id} = person;

    return { 
      name, 
      character, 
      headshot: `${imageUrl}${profile_path}`, 
      id: cast_id 
    };
  });

  return { director, writer, cast };
};