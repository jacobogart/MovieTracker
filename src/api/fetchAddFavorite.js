import { serverUrl } from "./pathNames";

export const fetchAddFavorite = (cardData) => {
  const url = `${serverUrl}/favorites/new`;
  const { 
    user, 
    title, 
    overview, 
    posterImg, 
    id, 
    rating, 
    releaseDate 
  } = cardData;
  let poster_path = posterImg.split('500')[1];
  const body = {
    movie_id: id, 
    user_id: user.id,
    title, 
    poster_path,
    release_date: releaseDate, 
    vote_average: rating, 
    overview
  };
  const init = {
    method: 'POST',
    body: JSON.stringify(body),
    headers: { 'Content-Type': 'application/json' }
  };

  return fetch(url, init)
    .then(response => {
      if (!response.ok) {
        throw Error('Failed to add favorite');
      } else {
        return response.json();
      }
    });
};
