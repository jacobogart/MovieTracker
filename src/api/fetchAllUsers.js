import { serverUrl } from "./pathNames";

export const fetchAllUsers = () => {
  return fetch(serverUrl)
    .then(response => {
      if (!response.ok) {
        throw new Error("Failed to create account");
      } else {
        return response.json();
      }
    })
    .then(data => data.data);
};
