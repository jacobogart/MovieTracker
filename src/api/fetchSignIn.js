export const fetchSignIn = (email, password) => {
  const body = JSON.stringify({ email, password })
  const init = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body 
  }
  const request = new Request("http://localhost:3000/api/users", init);
  return fetch(request)
    .then(response => response.json())
    .then(data => data.data)
}