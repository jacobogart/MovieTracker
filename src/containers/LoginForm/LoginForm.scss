.LoginForm {
  @include flex-column;
  @include form-layout;

  .submit-btn {
    background: rgba($accent, .9);

    &:hover {
      background: $accent;
    }
  }
}

.error-message {
  color: $accent;
  font-size: .94em;
  height: 20px;
}

.signup-link {
  font-style: italic;
  font-weight: 400;
  margin: 0 0 0 5px;
  text-decoration: underline;
}

.signup-link:hover {
  color: $accent;
}