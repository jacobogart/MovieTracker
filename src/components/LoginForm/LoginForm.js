import React, { Component } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { fetchSignIn } from '../../api/fetchSignIn';
import { updateUser } from '../../actions'

export class LoginForm extends Component {
  constructor() {
    super() 
    this.state = {
      username: '',
      password: ''
    }
  }
  handleSubmit = (e) => {
    e.preventDefault();
    const { username, password } = this.state;
    fetchSignIn(username, password)
      .then(user => this.props.updateUser(user.id, user.name))
    this.setState({
      username: '',
      password: ''
    })
  }

  handleChange = (e) => {
    this.setState({ [e.target.id]: e.target.value });
  } 

  render() {
    return (
      <form
        onSubmit={this.handleSubmit}
      >
        <label htmlFor="username">Username</label>
        <input 
          type="text" 
          id="username"
          value={this.state.username} 
          onChange={this.handleChange}
        />
        <label htmlFor="password">Password</label>
        <input 
          type="password" 
          id="password" 
          value={this.state.password}
          onChange={this.handleChange}
        />
        <button type='submit'>Submit</button>
        <p>
          Not a member? <Link to='/signup'>Create an account</Link>
        </p>
      </form>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  updateUser: (id, name) => dispatch(updateUser(id, name))
});

export default connect(null, mapDispatchToProps)(LoginForm);

