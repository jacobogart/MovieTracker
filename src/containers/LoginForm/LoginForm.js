import React, { Component } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { fetchSignIn } from '../../api/fetchSignIn';
import { updateUser } from '../../actions';

export class LoginForm extends Component {
  constructor() {
    super();
    this.state = {
      email: '',
      password: '',
      error: false
    }
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { email, password } = this.state;
    fetchSignIn(email, password)
      .then(user => {
        this.props.updateUser(user.id, user.name);
        this.props.hideLogin();
      })
      .catch(error => {
        this.setState({
          error: true,
          password: ""
        });
      });
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  } 

  render() {
    return (
      <form
        onSubmit={this.handleSubmit}
        className="LoginForm"
      >
        <label htmlFor="email-input">Email</label>
        <input 
          type="text" 
          id="email-input"
          name="email"
          value={this.state.email}
          onChange={this.handleChange}
        />
        <label htmlFor="password-input">Password</label>
        <input 
          type="password" 
          id="password-input"
          name="password" 
          value={this.state.password}
          onChange={this.handleChange}
        />
        {this.state.error && 
          <p className='error-message'>Incorrect password. Please try again.</p>
        }
        <button type="submit" className="submit-btn">Submit</button>
        <p>
          Not a member? <Link to="/signup" className="signup-link" onClick={this.props.hideLogin}>Create an account</Link>
        </p>
      </form>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  updateUser: (id, name) => dispatch(updateUser(id, name))
});

export default connect(null, mapDispatchToProps)(LoginForm);

