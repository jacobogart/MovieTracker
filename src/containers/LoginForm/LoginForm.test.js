import React from "react";
import { shallow } from "enzyme";
import { LoginForm, mapDispatchToProps } from "./LoginForm";
import { fetchSignIn } from "../../api/fetchSignIn";
import { updateUser } from "../../actions";

jest.mock("../../api/fetchSignIn");

const mockChangeEvent = {
  target: {
    name: "email",
    value: "email@email.com"
  }
};

const mockSubmitEvent = {
  preventDefault: () => {}
};

describe("LoginForm", () => {
  let wrapper, instance;
  let mockUpdateUser = jest.fn();
  let mockHideLogin = jest.fn();
  fetchSignIn.mockImplementation(() => Promise.resolve(1));

  beforeEach(() => {
    wrapper = shallow(
    <LoginForm 
      updateUser={mockUpdateUser} 
      hideLogin={mockHideLogin}
    />);
    instance = wrapper.instance();
  });

  afterEach(() => {
    fetchSignIn.mockClear();
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should have a defult state", () => {
    const mockDefaultState = {
      email: "",
      password: "",
      error: false
    };
    wrapper = shallow(<LoginForm />, { disableLifecycleMethods: true });
    expect(wrapper.state()).toEqual(mockDefaultState);
  });

  describe('Event Handlers', () => {
    it('should invoke handleChange on input change', () => {
      wrapper.find('#email-input').simulate('change', mockChangeEvent);
      expect(wrapper.state("email")).toEqual("email@email.com");
    });

    it('should invoke handleSubmit on form submit', () => {
      wrapper.find(".LoginForm").simulate('submit', mockSubmitEvent);
      expect(fetchSignIn).toHaveBeenCalled();
    })
  });

  describe("handleChange", () => {
    it("should set the appropriate state to the input value", () => {
      instance.handleChange(mockChangeEvent);
      expect(wrapper.state("email")).toEqual("email@email.com");
    });
  });

  describe("handleSubmit", () => {
    it("should invoke fetchSignIn", () => {
      instance.handleSubmit();
      expect(fetchSignIn).toHaveBeenCalled();
    });

    it("should invoke updateUser", () => {
      instance.handleSubmit(mockSubmitEvent);
      expect(mockUpdateUser).toHaveBeenCalled();
    });

    it("should invoke hideLogin", () => {
      instance.handleSubmit(mockSubmitEvent);
      expect(mockHideLogin).toHaveBeenCalled();
    });

    it('should set error in state to true if fetch fails', async () => {
      fetchSignIn.mockImplementation(() => Promise.reject());
      await instance.handleSubmit(mockSubmitEvent);
      expect(wrapper.state('error')).toEqual(true);
    })
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when using a function from MDTP', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = updateUser(1, 'Jacob');
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.updateUser(1, 'Jacob');
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
