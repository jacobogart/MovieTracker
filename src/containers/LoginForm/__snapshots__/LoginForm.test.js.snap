// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match the snapshot 1`] = `
<form
  className="LoginForm"
  onSubmit={[Function]}
>
  <label
    htmlFor="email-input"
  >
    Email
  </label>
  <input
    id="email-input"
    name="email"
    onChange={[Function]}
    type="email"
    value=""
  />
  <label
    htmlFor="password-input"
  >
    Password
  </label>
  <input
    id="password-input"
    name="password"
    onChange={[Function]}
    type="password"
    value=""
  />
  <p
    className="error-message"
  />
  <button
    className="submit-btn"
    type="submit"
  >
    Submit
  </button>
  <p>
    Not a member?
    <Link
      className="signup-link"
      onClick={[MockFunction]}
      to="/signup"
    >
      Create an account.
    </Link>
  </p>
</form>
`;
